modules/apps/portal-remote/portal-remote-cors-impl/src/main/java/com/liferay/portal/remote/cors/internal/BaseTrieURLToCORSSupportMapper.java

```diff
@@ -0,0 +1,81 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.portal.remote.cors.internal;
+
+import com.liferay.portal.kernel.util.Validator;
+
+/**
+ * @author Carlos Sierra Andrés
+ * @author Arthur Chan
+ */
+public abstract class BaseTrieURLToCORSSupportMapper
+	extends BaseURLToCORSSupportMapper {
+
+	@Override
+	public CORSSupport get(String urlPath) {
+		CORSSupport corsSupport = getWildcardCORSSupport(urlPath);
+
+		if (corsSupport != null) {
+			return corsSupport;
+		}
+
+		return getExtensionCORSSupport(urlPath);
+	}
+
+	protected abstract CORSSupport getExtensionCORSSupport(String urlPath);
+
+	protected abstract CORSSupport getWildcardCORSSupport(String urlPath);
+
+	protected void put(CORSSupport corsSupport, String urlPattern)
+		throws IllegalArgumentException {
+
+		if (corsSupport == null) {
+			throw new IllegalArgumentException("CORS support is null");
+		}
+
+		if (Validator.isBlank(urlPattern)) {
+			throw new IllegalArgumentException("URL pattern is blank");
+		}
+
+		try {
+			if (isWildcardURLPattern(urlPattern)) {
+				put(corsSupport, urlPattern, true);
+
+				return;
+			}
+
+			if (isExtensionURLPattern(urlPattern)) {
+				put(corsSupport, urlPattern, false);
+
+				return;
+			}
+
+			put(corsSupport, urlPattern, true);
+		}
+		catch (IndexOutOfBoundsException indexOutOfBoundsException) {
+			throw new IllegalArgumentException(
+				"URL pattern contains invalid characters",
+				indexOutOfBoundsException);
+		}
+	}
+
+	protected abstract void put(
```

Comment:
I am not sure if we shouldn't change this `put` method name to disambiguate it from the public one. Something like `triePut` or `putTrie`... I don't know

Reply:
Last time, I did something similar, calling a method like doPut(), Brian changed it back to just put()...

Reply:
yeah... I saw that... it was the same situation with both methods protected? Anyway... maybe we can keep it like it is and point to Brian that we are not sure how to name this because there have been previous renames and let him decide.